The streaming output/writter should be closed if the request connection is closed

Jersey is still trying to send a response after the client has closed the connection. This happens when a request is sent to download file data (without authorization), thus the client receives a 401 and closes the connection, but the jersey is still trying to write output.

``` Sep 18, 2019 2:48:05 PM org.glassfish.jersey.server.ServerRuntime$Responder writeResponse
radar-upload-connect-backend_1   | SEVERE: An I/O error has occurred while writing a response message entity to the container output stream.
radar-upload-connect-backend_1   | org.glassfish.jersey.server.internal.process.MappableException: java.io.IOException: Connection is closed
radar-upload-connect-backend_1   | 	at org.glassfish.jersey.server.internal.MappableExceptionWrapperInterceptor.aroundWriteTo(MappableExceptionWrapperInterceptor.java:67)
radar-upload-connect-backend_1   | 	at org.glassfish.jersey.message.internal.WriterInterceptorExecutor.proceed(WriterInterceptorExecutor.java:139)
radar-upload-connect-backend_1   | 	at org.glassfish.jersey.message.internal.MessageBodyFactory.writeTo(MessageBodyFactory.java:1111)
radar-upload-connect-backend_1   | 	at org.glassfish.jersey.server.ServerRuntime$Responder.writeResponse(ServerRuntime.java:638)
radar-upload-connect-backend_1   | 	at org.glassfish.jersey.server.ServerRuntime$Responder.processResponse(ServerRuntime.java:371)
radar-upload-connect-backend_1   | 	at org.glassfish.jersey.server.ServerRuntime$Responder.process(ServerRuntime.java:361)
radar-upload-connect-backend_1   | 	at org.glassfish.jersey.server.ServerRuntime$1.run(ServerRuntime.java:256)
radar-upload-connect-backend_1   | 	at org.glassfish.jersey.internal.Errors$1.call(Errors.java:248)
radar-upload-connect-backend_1   | 	at org.glassfish.jersey.internal.Errors$1.call(Errors.java:244)
radar-upload-connect-backend_1   | 	at org.glassfish.jersey.internal.Errors.process(Errors.java:292)
radar-upload-connect-backend_1   | 	at org.glassfish.jersey.internal.Errors.process(Errors.java:274)
radar-upload-connect-backend_1   | 	at org.glassfish.jersey.internal.Errors.process(Errors.java:244)
radar-upload-connect-backend_1   | 	at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:265)
radar-upload-connect-backend_1   | 	at org.glassfish.jersey.server.ServerRuntime.process(ServerRuntime.java:232)
radar-upload-connect-backend_1   | 	at org.glassfish.jersey.server.ApplicationHandler.handle(ApplicationHandler.java:679)
radar-upload-connect-backend_1   | 	at org.glassfish.jersey.grizzly2.httpserver.GrizzlyHttpContainer.service(GrizzlyHttpContainer.java:353)
radar-upload-connect-backend_1   | 	at org.glassfish.grizzly.http.server.HttpHandler$1.run(HttpHandler.java:200)
radar-upload-connect-backend_1   | 	at org.glassfish.grizzly.threadpool.AbstractThreadPool$Worker.doWork(AbstractThreadPool.java:569)
radar-upload-connect-backend_1   | 	at org.glassfish.grizzly.threadpool.AbstractThreadPool$Worker.run(AbstractThreadPool.java:549)
radar-upload-connect-backend_1   | 	at java.base/java.lang.Thread.run(Thread.java:835)
radar-upload-connect-backend_1   | Caused by: java.io.IOException: Connection is closed
radar-upload-connect-backend_1   | 	at org.glassfish.grizzly.nio.NIOConnection.assertOpen(NIOConnection.java:441)
radar-upload-connect-backend_1   | 	at org.glassfish.grizzly.http.io.OutputBuffer.write(OutputBuffer.java:663)
radar-upload-connect-backend_1   | 	at org.glassfish.grizzly.http.server.NIOOutputStreamImpl.write(NIOOutputStreamImpl.java:59)
radar-upload-connect-backend_1   | 	at org.glassfish.jersey.message.internal.CommittingOutputStream.write(CommittingOutputStream.java:200)
radar-upload-connect-backend_1   | 	at org.glassfish.jersey.message.internal.WriterInterceptorExecutor$UnCloseableOutputStream.write(WriterInterceptorExecutor.java:276)
radar-upload-connect-backend_1   | 	at kotlin.io.ByteStreamsKt.copyTo(IOStreams.kt:108)
radar-upload-connect-backend_1   | 	at kotlin.io.ByteStreamsKt.copyTo$default(IOStreams.kt:103)
radar-upload-connect-backend_1   | 	at org.radarbase.upload.resource.RecordResource$getContents$streamingOutput$1.write(RecordResource.kt:226)
radar-upload-connect-backend_1   | 	at org.glassfish.jersey.message.internal.StreamingOutputProvider.writeTo(StreamingOutputProvider.java:55)
radar-upload-connect-backend_1   | 	at org.glassfish.jersey.message.internal.StreamingOutputProvider.writeTo(StreamingOutputProvider.java:37)
radar-upload-connect-backend_1   | 	at org.glassfish.jersey.message.internal.WriterInterceptorExecutor$TerminalWriterInterceptor.invokeWriteTo(WriterInterceptorExecutor.java:242)
radar-upload-connect-backend_1   | 	at org.glassfish.jersey.message.internal.WriterInterceptorExecutor$TerminalWriterInterceptor.aroundWriteTo(WriterInterceptorExecutor.java:227)
radar-upload-connect-backend_1   | 	at org.glassfish.jersey.message.internal.WriterInterceptorExecutor.proceed(WriterInterceptorExecutor.java:139)
radar-upload-connect-backend_1   | 	at org.glassfish.jersey.server.internal.JsonWithPaddingInterceptor.aroundWriteTo(JsonWithPaddingInterceptor.java:85)
radar-upload-connect-backend_1   | 	at org.glassfish.jersey.message.internal.WriterInterceptorExecutor.proceed(WriterInterceptorExecutor.java:139)
radar-upload-connect-backend_1   | 	at org.glassfish.jersey.server.internal.MappableExceptionWrapperInterceptor.aroundWriteTo(MappableExceptionWrapperInterceptor.java:61)
radar-upload-connect-backend_1   | 	... 19 more
radar-upload-connect-backend_1   | Caused by: java.io.IOException: Broken pipe
radar-upload-connect-backend_1   | 	at java.base/sun.nio.ch.FileDispatcherImpl.write0(Native Method)
radar-upload-connect-backend_1   | 	at java.base/sun.nio.ch.SocketDispatcher.write(SocketDispatcher.java:47)
radar-upload-connect-backend_1   | 	at java.base/sun.nio.ch.IOUtil.writeFromNativeBuffer(IOUtil.java:113)
radar-upload-connect-backend_1   | 	at java.base/sun.nio.ch.IOUtil.write(IOUtil.java:58)
radar-upload-connect-backend_1   | 	at java.base/sun.nio.ch.IOUtil.write(IOUtil.java:50)
radar-upload-connect-backend_1   | 	at java.base/sun.nio.ch.SocketChannelImpl.write(SocketChannelImpl.java:466)
radar-upload-connect-backend_1   | 	at org.glassfish.grizzly.nio.transport.TCPNIOUtils.flushByteBuffer(TCPNIOUtils.java:125)
radar-upload-connect-backend_1   | 	at org.glassfish.grizzly.nio.transport.TCPNIOUtils.writeCompositeBuffer(TCPNIOUtils.java:64)
radar-upload-connect-backend_1   | 	at org.glassfish.grizzly.nio.transport.TCPNIOAsyncQueueWriter.write0(TCPNIOAsyncQueueWriter.java:105)
radar-upload-connect-backend_1   | 	at org.glassfish.grizzly.nio.transport.TCPNIOAsyncQueueWriter.write0(TCPNIOAsyncQueueWriter.java:82)
radar-upload-connect-backend_1   | 	at org.glassfish.grizzly.nio.AbstractNIOAsyncQueueWriter.write(AbstractNIOAsyncQueueWriter.java:236)
radar-upload-connect-backend_1   | 	at org.glassfish.grizzly.nio.AbstractNIOAsyncQueueWriter.write(AbstractNIOAsyncQueueWriter.java:145)
radar-upload-connect-backend_1   | 	at org.glassfish.grizzly.nio.AbstractNIOAsyncQueueWriter.write(AbstractNIOAsyncQueueWriter.java:47)
radar-upload-connect-backend_1   | 	at org.glassfish.grizzly.nio.transport.TCPNIOTransportFilter.handleWrite(TCPNIOTransportFilter.java:102)
radar-upload-connect-backend_1   | 	at org.glassfish.grizzly.filterchain.TransportFilter.handleWrite(TransportFilter.java:167)
radar-upload-connect-backend_1   | 	at org.glassfish.grizzly.filterchain.ExecutorResolver$8.execute(ExecutorResolver.java:87)
radar-upload-connect-backend_1   | 	at org.glassfish.grizzly.filterchain.DefaultFilterChain.executeFilter(DefaultFilterChain.java:260)
radar-upload-connect-backend_1   | 	at org.glassfish.grizzly.filterchain.DefaultFilterChain.executeChainPart(DefaultFilterChain.java:177)
radar-upload-connect-backend_1   | 	at org.glassfish.grizzly.filterchain.DefaultFilterChain.execute(DefaultFilterChain.java:109)
radar-upload-connect-backend_1   | 	at org.glassfish.grizzly.filterchain.DefaultFilterChain.process(DefaultFilterChain.java:88)
radar-upload-connect-backend_1   | 	at org.glassfish.grizzly.ProcessorExecutor.execute(ProcessorExecutor.java:53)
radar-upload-connect-backend_1   | 	at org.glassfish.grizzly.filterchain.FilterChainContext.write(FilterChainContext.java:866)
radar-upload-connect-backend_1   | 	at org.glassfish.grizzly.filterchain.FilterChainContext.write(FilterChainContext.java:834)
radar-upload-connect-backend_1   | 	at org.glassfish.grizzly.http.io.OutputBuffer.flushBuffer(OutputBuffer.java:1068)
radar-upload-connect-backend_1   | 	at org.glassfish.grizzly.http.io.OutputBuffer.write(OutputBuffer.java:695)
radar-upload-connect-backend_1   | 	... 33 more
```

--- a/radar-upload-backend/src/main/java/org/radarbase/upload/resource/RecordResource.kt
+++ b/radar-upload-backend/src/main/java/org/radarbase/upload/resource/RecordResource.kt
@@ -33,6 +33,7 @@ import org.radarbase.upload.exception.ConflictException
 import org.radarcns.auth.authorization.Permission
 import org.radarcns.auth.authorization.Permission.*
 import org.slf4j.LoggerFactory
+import java.io.IOException
 import java.io.InputStream
 import java.net.URI
 import javax.annotation.Resource
@@ -222,10 +223,16 @@ class RecordResource {
         logger.debug("Reading record $recordId file $fileName of size ${recordContent.size}")
 
         val streamingOutput = StreamingOutput { out ->
-            recordRepository.readFileContent(recordId, record.metadata.revision, recordContent.fileName).use {
-                it?.stream?.copyTo(out) ?: throw RbNotFoundException("file_not_found", "Cannot read content of file $fileName from record $recordId")
+            try {
+                recordRepository.readFileContent(recordId, record.metadata.revision, recordContent.fileName).use {
+                    it?.stream?.copyTo(out)
+                            ?: throw RbNotFoundException("file_not_found", "Cannot read content of file $fileName from record $recordId")
+                }
+                out.flush()
+            } catch (ex: IOException) {
+                logger.error("Failed to respond with file contents: {}. Caused by?: {}",
+                        ex.toString(), ex.cause?.toString())
             }
-            out.flush()
         }
 
         return Response
@@ -277,20 +284,28 @@ class RecordResource {
     fun getRecordLogs(
             @PathParam("recordId") recordId: Long): Response {
         val record = ensureRecord(recordId, SUBJECT_READ)
+        recordRepository.readLogs(recordId)
+                ?: throw RbNotFoundException("log_not_found", "Cannot find logs for record with record id $recordId")
 
         val streamingOutput = StreamingOutput { out ->
-            recordRepository.readLogContents(recordId).use { clobReader ->
-                clobReader ?: throw RbNotFoundException("log_not_found", "Cannot find logs for record with record id $recordId")
+            try {
+                recordRepository.readLogContents(recordId).use { clobReader ->
+                    clobReader
+                            ?: throw RbNotFoundException("log_not_found", "Cannot find logs for record with record id $recordId")
 
-                clobReader.stream.use { reader ->
-                    val buffer = CharArray(LOG_BUFFER_SIZE)
+                    clobReader.stream.use { reader ->
+                        val buffer = CharArray(LOG_BUFFER_SIZE)
 
-                    generateSequence { reader.read(buffer) }
-                            .takeWhile { n -> n != -1 }
-                            .forEach { n -> out.write(String(buffer, 0, n).toByteArray()) }
+                        generateSequence { reader.read(buffer) }
+                                .takeWhile { n -> n != -1 }
+                                .forEach { n -> out.write(String(buffer, 0, n).toByteArray()) }
+                    }
                 }
+                out.flush()
+            } catch (ex: IOException) {
+                logger.error("Failed to respond with log contents: {}. Caused by?: {}",
+                        ex.toString(), ex.cause?.toString())
             }
-            out.flush()
         }
 
         return Response.ok(streamingOutput, "text/plain")

